client/engine/messageservice/simple-tcp/simpletcp-messageservice.go

```diff
@@ -0,0 +1,154 @@
+package simpletcp
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+const (
+	CONN_TYPE = "tcp"
+	DELIMETER = '\n'
+)
+
+// SimpleTCPMessageService is a rudimentary message service that uses TCP to send and receive messages
+type SimpleTCPMessageService struct {
+	in  chan protocols.Message // for receiving messages from engine
+	out chan protocols.Message // for sending message to engine
+
+	peers map[types.Address]string
+
+	listener net.Listener // The listener for incoming connections on our port
+
+	quit chan struct{} // quit is used to signal the goroutine to stop
+
+}
+
+// NewTestMessageService returns a running SimpleTcpMessageService listening on the given url
+func NewSimpleTCPMessageService(myUrl string, peers map[types.Address]string) *SimpleTCPMessageService {
+
+	l, err := net.Listen(CONN_TYPE, myUrl)
+	if err != nil {
+		panic(err)
+	}
+	h := &SimpleTCPMessageService{
+		in:    make(chan protocols.Message, 5),
+		out:   make(chan protocols.Message, 5),
+		peers: peers,
+
+		listener: l,
+		quit:     make(chan struct{}),
+	}
+
+	go h.listenForIncoming()
+	go h.listenForOutgoing()
+
+	return h
+
+}
+
+// listenForOutgoing listens for any messages that should be sent out to peers and sends them
+func (s *SimpleTCPMessageService) listenForOutgoing() {
+	for {
+		select {
+		case <-s.quit:
+			return
+
+		case m := <-s.in:
+			{
+				peer, ok := s.peers[m.To]
+
+				if !ok {
+					panic(fmt.Errorf("no peer port for %s", m.To))
+				}
+
+				raw, err := m.Serialize()
+
+				// Append the delimiter to the message to indicate the end of the message
+				raw = fmt.Sprintf("%s%c", raw, DELIMETER)
+
+				if err != nil {
+					panic(err)
+				}
+
+				conn, err := net.Dial(CONN_TYPE, peer)
+				if err != nil {
+					select {
+					case <-s.quit: // If we are quitting we can ignore the error
```

Comment:
This pattern (checking first whether a quit signal has been sent, and if it hasn't, checking _again_ if we hit an error condition) is repeated a lot -- is it unavoidable, do you think?

Reply:
I've tried to clean this up in https://github.com/statechannels/go-nitro/pull/681/commits/2c6f02021a4d7073037036fbc77f49bebd22cc6f. I think anytime we're dealing with the `listener` we have to deal with errors that might arise from calling `s.listener.Close()`